src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java

```diff
@@ -1,61 +1,61 @@
 package org.broadinstitute.hellbender.utils.pileup;
 
-import com.google.api.services.genomics.model.Read;
 import com.google.common.annotations.VisibleForTesting;
 import htsjdk.samtools.SAMFileHeader;
 import htsjdk.samtools.util.Locatable;
-import org.broadinstitute.hellbender.exceptions.GATKException;
 import org.broadinstitute.hellbender.exceptions.UserException;
-import org.broadinstitute.hellbender.tools.walkers.qc.Pileup;
 import org.broadinstitute.hellbender.utils.BaseUtils;
 import org.broadinstitute.hellbender.utils.QualityUtils;
 import org.broadinstitute.hellbender.utils.Utils;
-import org.broadinstitute.hellbender.utils.fragments.FragmentCollection;
 import org.broadinstitute.hellbender.utils.read.GATKRead;
-import org.broadinstitute.hellbender.utils.read.ReadUtils;
-import org.broadinstitute.hellbender.engine.AlignmentContext;
 
 import java.util.*;
 import java.util.function.Predicate;
 import java.util.function.ToIntFunction;
 import java.util.stream.Collectors;
-import java.util.stream.Stream;
 
 /**
  * Represents a pileup of reads at a given position.
  */
 public class ReadPileup implements Iterable<PileupElement> {
     private final Locatable loc;
-    private final List<PileupElement> pileupElements;
-
-    /** Constant used by samtools to downgrade a quality for overlapping reads that disagrees in their base. */
-    public static final double SAMTOOLS_OVERLAP_LOW_CONFIDENCE = 0.8;
+    private PileupElementTracker pileupElements;
 
     /**
      * Create a new pileup at loc, using the reads and their corresponding
      * offsets.
      * Note: This constructor keeps an alias to the given list.
      */
     public ReadPileup(final Locatable loc, final List<PileupElement> pileup) {
+        this(loc, pileup, false);
+    }
+
+    /**
+     * Create a new pileup at loc, using the reads and their corresponding
+     * offsets.
+     * Note: This constructor keeps an alias to the given list.
+     */
+    public ReadPileup(final Locatable loc, final List<PileupElement> pileup, final boolean preSorted) {
         Utils.nonNull(loc, "loc is null");
         Utils.nonNull(pileup, "element list is null");
         this.loc = loc;
-        this.pileupElements = pileup;
+        this.pileupElements = new UnifiedPileupElementTracker(pileup, preSorted);
     }
 
     /**
-     * Create a new pileup at loc, using an stratified pileup
-     * Note: the current implementation of ReadPileup does not efficiently retrieve the stratified pileup
+     * Create a new pileup at loc, using a stratified pileup.
      */
     public ReadPileup(final Locatable loc, final Map<String, ReadPileup> stratifiedPileup) {
-        this(loc, stratifiedPileup.values().stream().flatMap(ReadPileup::getElementStream).collect(Collectors.toList()));
+        this.loc = loc;
+        this.pileupElements = new PerSamplePileupElementTracker(stratifiedPileup.entrySet().stream()
```

Comment:
Self note: this was changed in #4031 for a performance improvement for stratified pileups. We should check if there is any regression...